### DISCUSSION TITLE
Frustrating development setup

### DISCUSSION THREAD

**Discussion by rchl** [vuejs/language-tools#4977] [Q&A]
While I did manage to set development workspace for vue-language-tools a couple of times before, after I come back to it, it typically just breaks and forces me to figure it all out again which makes it very cumbersome and annoying to contribute to this project.

Can someone who is active on the project describe the exact steps to set up development workspace to be able to develop vue-language-tools?

Like:
1. Clone https://github.com/volarjs/workspace
2. Run `pnpm i`
3. Run `pnpm run setup`
4. Run `pnpm watch:vue`
> typically it throw a bunch of errors at this point
5. Run `cd vue && pnpm i && pnpm build`
> now it compiles successfully but changes done in `/services` do NOT get picked up when compiling vue language server in `/vue`. I think it doesn't have proper package linking set up and instead uses released packages from `npm`.

How to make packages in `/vue` depend on packages from `/services` rather than upstream ones?

I feel like previously I've been deleting `pnpm-lock.yml` for things to get linked properly but that caused more errors to to random packages changing versions.

@johnsoncodehk @so1ve @KazariEX

**Comment by KazariEX** [comment]
We need to run `tsc` in `extensions/vscode` first.

**Reply by rchl** [reply]
`npx tsc` or `pnpm build`?

But that anyway goes a bit further from what I was asking. I'm not even asking anything vscode-specific. I'm just wondering how to get packages to link properly so that I can build `@vue/language-server` and take advantage of linked pages.

**Reply by KazariEX** [reply]
It seems that this "workspace" is simply cloning multiple repositories into one folder, but since it's a monorepo perhaps we can manually modify the path of packages that need to be linked in `package.json` to achieve it?

**Reply by KazariEX** [reply]
Sorry, I found that path has been overridden by pnpm, it should be linked correctly.

**Reply by rchl** [reply]
Yeah, the links are set up in https://github.com/volarjs/workspace/blob/1acac605768346a0ee63426ca86de5a369fabc39/package.json#L23-L50 but I just don't see it working consistently...

**Reply by KazariEX** [reply]
Could you try running `watch` instead of `watch:base`?

**Reply by rchl** [reply]
Not helping.

Just simply following my posted steps, at step 4 a ton of errors is thrown where the first one is:

```
../services/packages/emmet/index.ts:3:39 - error TS2307: Cannot find module 'EmmetFlatNode' or its corresponding type declarations.

3 import type { Node, Stylesheet } from 'EmmetFlatNode';
                                        ~~~~~~~~~~~~~~~
```

So it seems to me that the linking works at this point since it references /services but it can't find the [EmmetFlatNode module declaration](https://github.com/volarjs/services/blob/79afca2f3b2c8e3a5cf867252b18002bbde4bca0/packages/emmet/lib/typings/EmmetFlatNode.d.ts#L6-L6) for some reason

I think that running `pnpm i` in sub-directories of the workspace just removes the linking and makes those use released dependencies which makes it compile but doesn't fit my use case where I want to edit stuff in `/services` and include those in compiled `@vue/language-server`.

**Reply by rchl** [reply]
@johnsoncodehk do you have any input here? What is the flow to follow for someone who wants to develop the language server (ignore vscode part if it makes it simpler).
